### ISSUE TITLE
docker: Support the postgres initdb.d startup protocol

### ISSUE THREAD

**Issue by quodlibetor** [MaterializeInc/materialize#4738]
The [postgres docker image](https://hub.docker.com/_/postgres) supports very convenient startup via an initdb.d directory:

> If you would like to do additional initialization in an image derived from this one, add one or more *.sql, *.sql.gz, or *.sh scripts under /docker-entrypoint-initdb.d (creating the directory if necessary). After the entrypoint calls initdb to create the default postgres user and database, it will run any *.sql files, run any executable *.sh scripts, and source any non-executable *.sh scripts found in that directory to do further initialization before starting the service.

It would be nice if we supported at least a limited subset of that, although it will complicate our startup somewhat.

**Comment by benesch** [comment]
A few notes:

  * The materialized container doesn't presently bundle `psql`, which means there's no easy way to just whip this up in the Docker container (i.e., without changing materialized), unless we want to start bundling `psql` in the container.
  * We used to have a `--bootstrap-sql` option to Materialize that provided this behavior, but it was removed because its semantics were confusing. Empirically, a lot of users don't have the right mental model for "this option only applies if I'm creating a new mzdata directory." Making it an error to specify `--bootstrap-sql` if the catalog is not being bootstrapped is also empirically problematic, because a lot of tools (e.g. Docker Compose) can't handle that complexity.
  * The Postgres design has some caveats:
    > One common problem is that if one of your /docker-entrypoint-initdb.d scripts fails (which will cause the entrypoint script to exit) and your orchestrator restarts the container with the already initialized data directory, it will not continue on with your scripts.

I think the product ask here is "make it easier to install initial catalog items in Materialize." That's something I'm very much on board with! What I'm wondering is whether the Docker container is the optimal place to provide that feature. If folks are going to be using k8s, is there some k8s feature we should point people towards that makes this trivial?
